dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramInstanceEventSupplier.java

```diff
@@ -0,0 +1,306 @@
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+/**
+ * @author Luciano Fiandesio
+ */
+@Component( "workContextProgramInstancesEventSupplier" )
+public class ProgramInstanceEventSupplier extends AbstractSupplier
+{
+    private final ProgramEventSupplier programSupplier;
+
+    public ProgramInstanceEventSupplier( NamedParameterJdbcTemplate jdbcTemplate, ProgramEventSupplier programSupplier,
+        Environment environment )
+    {
+        super( jdbcTemplate, environment );
+        this.programSupplier = programSupplier;
+    }
+
+    public Map<String, ProgramInstance> get( ImportOptions importOptions,
+        Map<String, Pair<TrackedEntityInstance, Boolean>> teiMap, List<Event> events )
+    {
+        if ( events == null )
+        {
+            return new HashMap<>();
+        }
+
+        // Collect all the program instance UIDs to pass as SQL query argument
+        Set<String> programInstanceUids = events.stream()
+            .map( Event::getEnrollment )
+            .filter( StringUtils::isNotEmpty ).collect( Collectors.toSet() );
+
+        Map<String, ProgramInstance> programInstances = new HashMap<>();
+
+        if ( !programInstanceUids.isEmpty() )
+        {
+            // Create a bi-directional map enrollment uid -> event uid
+            Multimap<String, String> programInstanceToEvent = HashMultimap.create();
+            for ( Event event : events )
+            {
+                programInstanceToEvent.put( event.getEnrollment(), event.getUid() );
+            }
+
+            // Collect all the Program Stage Instances specified in the Events
+            // (enrollment
+            // property)
+            programInstances = getProgramInstancesByUid( importOptions, events, programInstanceToEvent,
+                programInstanceUids );
+        }
+
+        mapExistingEventsToProgramInstances( importOptions, events, programInstances );
+
+        mapEventsToProgramInstanceByTei( importOptions, events, programInstances, teiMap );
+
+        return programInstances;
+    }
+
+    /**
+     * Loop through the events and check if there is any event left without a
+     * Program Instance: for each Event without a PI, try to fetch the Program
+     * Instance by Program and Tracked Entity Instance
+     */
+    private void mapEventsToProgramInstanceByTei( ImportOptions importOptions, List<Event> events,
+        Map<String, ProgramInstance> programInstances, Map<String, Pair<TrackedEntityInstance, Boolean>> teiMap )
+    {
+        for ( Event event : events )
+        {
+            if ( !programInstances.containsKey( event.getUid() ) )
+            {
+                Pair<TrackedEntityInstance, Boolean> teiPair = teiMap.get( event.getUid() );
```

Comment:
TeiMap and TeiPair are not really explicative.
What does the boolean represent here?

Reply:
I have no idea what is this about, it is old code and I didn't tale the time to verify as this issue is about a complete different argument